src/main/java/com/conveyal/r5/profile/Propagater.java

```diff
@@ -0,0 +1,88 @@
+package com.conveyal.r5.profile;
+
+import com.conveyal.r5.streets.LinkedPointSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.function.ToIntFunction;
+import java.util.stream.IntStream;
+
+/**
+ * A propagater does propagation, like how an alligator does allegation.
+ */
+public class Propagater {
+    private static final Logger LOG = LoggerFactory.getLogger(Propagater.class);
+
+    /** Times at transit stops for each iteration */
+    public final int[][] travelTimesToStopsEachIteration;
+
+    /** Times at targets using the street network */
+    public final int[] nonTransferTravelTimesToTargets;
+
+    /** The linked targets */
+    public final LinkedPointSet targets;
+
+    /** the profilerequest (used for walk speed etc.) */
+    public final ProfileRequest request;
+
+    public Propagater (int[][] travelTimesToStopsEachIteration, int[] nonTransferTravelTimesToTargets, LinkedPointSet targets, ProfileRequest request) {
+        this.travelTimesToStopsEachIteration = travelTimesToStopsEachIteration;
+        this.nonTransferTravelTimesToTargets = nonTransferTravelTimesToTargets;
+        this.targets = targets;
+        this.request = request;
+    }
+
+    /**
+     * perform the propagation, and call the reducer with the travel time to each target for each iteration.
+     *
+     * For large networks, making a 2-D array of all the travel times to every target is too large to fit in memory.
+     */
+    public int[] propagate (ToIntFunction<int[]> reducer) {
+        long startTime = System.currentTimeMillis();
+        // propagate the transit times
+        // NB if this is slow, parallelize
+        // This can use something akin to range-RAPTOR where we only update the travel time if it changes from one
+        // iteration to the next, but that requires working on clock times, so we'd have to pass more information down
+        // into this function; I'm not sure it's worth it.
+        int[] result = new int[travelTimesToStopsEachIteration.length];
+        for (int iteration = 0; iteration < travelTimesToStopsEachIteration.length; iteration++) {
+            int[] travelTimesToTargetsThisIteration = Arrays.copyOf(nonTransferTravelTimesToTargets, nonTransferTravelTimesToTargets.length);
+            int[] travelTimesToStopsThisIteration = travelTimesToStopsEachIteration[iteration];
+
+            // avoid float math in loop below
```

Comment:
This is because we had this same loop in the previous propagation code and at some point determined it to be slow. It's probably not because of floating-point ops but rather casts, although I don't know if we ever tried it with single-precision floats.